src/events/provider.py

```diff
@@ -57,33 +57,38 @@ def _on_database_requested(self, event: DatabaseRequestedEvent) -> None:
 
         if not self.charm.unit.is_leader():
             return
+        try:
+            if not self.context.is_authentication_enabled():
+                raise NotImplementedError(
+                    "Authentication has not been enabled yet! "
+                    "Please integrate kyuubi-k8s with postgresql-k8s "
+                    "over auth-db relation endpoint."
+                )
 
-        if not self.context.is_authentication_enabled():
-            raise NotImplementedError(
-                "Authentication has not been enabled yet! "
-                "Please integrate kyuubi-k8s with postgresql-k8s "
-                "over auth-db relation endpoint."
+            auth = AuthenticationManager(self.context.auth_db)
+            service_manager = ServiceManager(
+                namespace=self.charm.model.name,
+                unit_name=self.charm.unit.name,
+                app_name=self.charm.app.name,
             )
-        auth = AuthenticationManager(self.context.auth_db)
-        service_manager = ServiceManager(
-            namespace=self.charm.model.name,
-            unit_name=self.charm.unit.name,
-            app_name=self.charm.app.name,
-        )
-        try:
+
             username = f"relation_id_{event.relation.id}"
             password = auth.generate_password()
             auth.create_user(username=username, password=password)
 
             kyuubi_endpoint = service_manager.get_service_endpoint(
                 expose_external=self.charm.config.expose_external
             )
-            jdbc_uri = f"jdbc:hive2://{kyuubi_endpoint}/" if kyuubi_endpoint else ""
+            jdbc_uri = (
+                f"jdbc:hive2://{kyuubi_endpoint.host}:{kyuubi_endpoint.port}/"
```

Comment:
Might be easier to add `__str__()` to `IPEndpoint` and `DNSEndpoint` and cast instead.

Reply:
Yes, absolutely. Added to our housekeeping todo list, thanks